### PR TITLE
feat: Include an SNS in the CI testnet

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
We need e2e tests for the SNS.  The first thing we need for that is a testnet setup that includes an SNS.

# Changes
- Run the existing e2e tests against a local dfx server that has SNS set up.
  - Note: This does NOT deploy multiple SNSs, open SNSs and so on.  We will need that for some tests.  This PR just deploys the setup needed to support SNSs, and the largest part of running SNSs is the setup.

# Tests
- There is no new functionality, so existing tests should suffice

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -64,16 +65,28 @@ jobs:
           cargo --version
           ic-cdk-optimizer --version
```
Comment (bitdivine) [inline .github/workflows/build.yml:66]: We use the snsdemo scripts to deploy an SNS.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -64,16 +65,28 @@ jobs:
...
+        with:
+          repository: 'dfinity/snsdemo'
+          path: 'snsdemo'
+          # Version from Mon Jan 16 10:10:47
+          ref: 'e1d7dd4c10fb8a731a534ae2ef4b08474fea5f6b'
+
+      - name: Add SNS scripts to the path
+        run: |
+          echo "$PWD/snsdemo/bin" >> $GITHUB_PATH
+
```
Comment (bitdivine) [inline .github/workflows/build.yml:79]: This installs things such as dfx, jq and idl2json that we probably already have, but just to be sure that we don't have strange breakages due to something as silly as this, it seems like a good idea to explicitly ask for the snsdemo tools.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -64,16 +65,28 @@ jobs:
...
+      - name: Install SNS script dependencies
         run: |
-          dfx start --background --clean
+          dfx-sns-demo-install
```
Comment (bitdivine) [inline .github/workflows/build.yml:83]: This is where the action happens.  SNS joins the party.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -100,10 +113,6 @@ jobs:
         run: |
           DFX_NETWORK=local ENV_OUTPUT_FILE=./e2e-tests/.env ./config.sh
```
Comment (bitdivine) [inline .github/workflows/build.yml:116]: I think we can throw this away.  Green CI suggests so too.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -64,16 +65,28 @@ jobs:
...
+          path: 'snsdemo'
+          # Version from Mon Jan 16 10:10:47
+          ref: 'e1d7dd4c10fb8a731a534ae2ef4b08474fea5f6b'
+
+      - name: Add SNS scripts to the path
+        run: |
+          echo "$PWD/snsdemo/bin" >> $GITHUB_PATH
+
+      - name: Install SNS script dependencies
         run: |
```
Comment (bitdivine) [inline .github/workflows/build.yml:81]: I have moved start-replica to the script that deploys the testnet.

File: scripts/dfx-nns-deploy-custom [hunk 6]
```diff
@@ -24,8 +25,24 @@ rm -f nns-dapp.wasm nns-dapp_local.wasm
 : "Get our chosen version of II and inject it into the wasm cache."
 "$SOURCE_DIR"/dfx-software-ii-cache
```
Comment (bitdivine) [inline scripts/dfx-nns-deploy-custom:27]: Maybe we want just the NNS sometimes.  If the SNS is not needed, deploying without it is significantly faster.  So I have kept the old behaviour along with the new, full fat network with SNS.

File: .github/workflows/build.yml (hunk 4 above)
Comment (lmuntaner) [reply .github/workflows/build.yml:116]: Do you mean the creation of the config file?

File: scripts/dfx-nns-deploy-custom [hunk 7]
```diff
@@ -24,8 +25,23 @@ rm -f nns-dapp.wasm nns-dapp_local.wasm
...
-dfx nns install
+case "$DFX_NETWORK_FEATURES" in
+nns)
+  : "Deploy nns-dapp with other NNS canisters"
+  dfx start --background --clean
+  dfx nns install
+  ;;
+sns)
+  : "Deploy nns-dapp with NNS and SNS canisters"
+  dfx-sns-demo
```
Comment (lmuntaner) [inline scripts/dfx-nns-deploy-custom:36]: Can we add a comment that this comes from the snsdemo repository?

**Review by lmuntaner — APPROVED** [review]
Looks good to me!

Just one comment, but already approving.

Sorry that I removed one of your comments. I wanted to remove mine 😅 

But I agree, let's remove the populate.

We can even clean it from `./deploy.sh` later as well.

File: scripts/dfx-nns-deploy-custom (hunk 7 above)
Comment (bitdivine) [reply scripts/dfx-nns-deploy-custom:36]: Done

**Pull request merged**